docs/strings.md

```diff
@@ -0,0 +1,464 @@
+
+Random access in a string is a complex concept.
+Random access in C strings or C++ `std::string` is easy because they are simply arrays of code units but it is also meaningless in the context of variable-length encodings (byte *n* in a UTF-8 string, for example, may not be the start of a valid code point, let alone a grapheme cluster).
+Most random access wants to, at a minimum, index on code points and often on grapheme clusters.
+
+# Modification
+
+For mutable strings, inserting, deleting, and replacing substrings often need to be fast.
+This typically leads to mutable strings using a twine or similar representation.
+Even in-place modification of a character or a substring with a string of the same length can involve changing the number of code units in a string if the characters are replaced with different-width ones.
+This provides a strong incentive to avoid mandating contiguous storage for strings and is one of the common reasons for avoiding `std::string` in C++.
+
+# Comparison
+
+There are four common kinds of comparison on strings:
+
+ - Canonicalisation-unaware equality.
+   Do these two strings contain the same set of Unicode code points?
+ - Canonicalisation-aware equality.
+   Do these two strings contain a sequence of Unicode code points that are equal under canonicalisation.
+ - Locale-unaware ordered comparison.
+   What is a stable ordering between these two strings that is fast to compute?
+ - Locale-aware collation.
+   What is the ordering between these two strings, with the comparison computed using the Unicode collation rules?
+
+The first and third of these depend only on the strings being able to expose length and iterators that export a common character encoding.
+The second and fourth require a complex unicode library and should probably be in a separate package.
+
+# URL and path operations
+
+Some frameworks provide operations on strings for handling path and URL data.
+These are increasingly being deprecated and being moved to specific path and URL types that enforce validity and use structured storage until they must be exported to a string.
+It is unlikely that a Verona string design will want to include any of these operations in the core string APIs.
+
+Encodings
+---------
+
+Unicode defines three common serialisations, with different properties:
+
+ - UTF-8 is the most space-efficient for most Latin alphabet languages (an Emoji) and encodes every Unicode code point in 1-4 bytes.
+   The variation in length can make processing and random access indexed by code point expensive.
+ - UTF-16 is the most space-efficient for CJK languages and encodes every Unicode code point to 2 or 4 bytes.
+   As with UTF-8, random access requires additional metadata, though this can typically be omitted for strings in European languages, where every non-Emoji character will be 2 bytes.
+ - UTF-32 is the least space-efficient encoding, using 4 bytes for every character, but has the advantage of being a fixed-length encoding.
```

Comment:
Maybe add "(despite fixed-length of codepoints, grapheme cluster indexing remains challenging)"?